core/headerchain.go

```diff
@@ -176,26 +184,21 @@ func (hc *HeaderChain) WriteHeader(header *types.Header) (status WriteStatus, er
 			headNumber = headHeader.Number.Uint64() - 1
 			headHeader = hc.GetHeader(headHash, headNumber)
 		}
+
+		if err := WriteHeadHeaderHash(hc.chainDb, hash); err != nil {
+			glog.Fatalf("failed to insert head header hash: %v", err)
+		}
```

Comment:
I'm not sure this block should be moved. The head header is the topmost index that signals the current position of the header chain. Imho it should not be updated to the new head before the canonical number is written out as the invariant is that if head points to X, everything until X is present and complete.